openvdb/openvdb/points/PointStatistics.h

```diff
@@ -0,0 +1,609 @@
+#include "../Types.h"
+#include "../math/Math.h"
+#include "../tree/LeafManager.h"
+
+#include <tbb/parallel_reduce.h>
+#include <tbb/parallel_for.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+/// @brief Evaluates the minimum and maximum active values of a point attribute.
+/// @details  This function sets the arguments min and max to the minimum and
+///   and maximum values of a given point attributes. The ValueType of the
+///   attribute must be copy constructible and support less than and greater
+///   than operators. The ValueType does not need to support zero initialization
+///   or define its own numerical limits. This method will throw only if the
+///   templated ValueType does not match the given attribute. This method will
+///   return true if min and max have been set, false otherwise. The function is
+///   deterministic.
+/// @note  The value type of the min/max calculations must match the value type
+///   of the attribute. For vectors and matrices, this results in per component
+///   comparisons. See evalExtents for magnitudes or more custom control.
+/// @param points     the point tree
+/// @param attribute  the attribute to reduce
+/// @param min        the computed min value
+/// @param max        the computed max value
+/// @param filter     a filter to apply to points
+/// @param minTree    if provided, builds a tiled tree of localised min results
+/// @param maxTree    if provided, builds a tiled tree of localised max results
+/// @return true if min and max have been set, false otherwise. Can be false if
+///   no points were processed or if the tree was empty.
+/// @warning if "P" is provided as the attributes, it is evaluated in voxel space
+template <typename ValueT,
+    typename CodecT = UnknownCodec,
+    typename FilterT = NullFilter,
+    typename PointDataTreeT = points::PointDataTree>
+inline bool evalMinMax(const PointDataTreeT& points,
+        const std::string& attribute,
+        ValueT& min,
+        ValueT& max,
+        const FilterT& filter = NullFilter(),
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* minTree = nullptr,
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* maxTree = nullptr);
+
+/// @brief Evaluates the average active value of a point attribute.
+/// @details  This function sets the argument average to the average value of a
+///   given point attributes. The ValueType of the attribute must be copy
+///   constructible, support the same type addition operator and the division
+///   operator from a floating point scalar. This method allows for a different
+///   type ResultT to be provided which is used to compute the final result
+///   as well as for intermediate arithmetic. This allows for higher precision
+///   results if overflow or underflow is a concern, as well as computing
+///   floating point results from integer types. This method will throw only if
```

Comment:
It's impossible to assume all use cases - I personally would find it strange that a reduction performed at a different (higher) precision than my requested `ValueType`. The same goes for integers, it may be that a point set has been configured such that its values and counts are divisible. These are perhaps edge cases - I wouldn't be completely opposed to removing this option but I'd like to hear if anyone else has thoughts.

Reply:
Note that this also supports a different destination type for custom types

Reply:
As discussed, I will be changing `average` to work at double precision for floats and integers and to always return a double. `accumulate` will work at the higher corresponding type precision and return that type.

Reply:
I've added new type traits classes to handle this. The thing is that I still wanted the `ResultT` type to be templated so that the function is extensible to future/custom types. At the moment both `average` and `accumulate` will change their precision based on the input result type:


```
template <typename ValueT,
    typename ResultT = typename ConvertType<ValueT, double>::Type,
     ....>
inline bool evalAverage(const PointDataTreeT& points,
        ResultT& average,
        ...);

double d;
evalAverage</*attr type*/float>(points, d); // runs at double precision

float f; 
evalAverage</*attr type*/float>(points, f); // runs at float precision
```

The alternative would be to do:

```
template <typename ValueT,
     ....>
inline bool evalAverage(const PointDataTreeT& points,
         typename ConvertType<ValueT, double>::Type& average,
        ...);

double d;
evalAverage</*attr type*/float>(points, d); // runs at double precision

// float f; 
// evalAverage</*attr type*/float>(points, f); // DOESN'T COMPILE
```

I feel this is pretty limiting. @danrbailey thoughts?

Reply:
This is now resolved for both average and total methods